src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionContextFactory.cs

```diff
@@ -0,0 +1,52 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.IO.Pipelines;
+using System.Net.Sockets;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets
+{
+    /// <summary>
+    /// A factory for socket based connections contexts.
+    /// </summary>
+    public sealed class SocketConnectionContextFactory : ISocketConnectionContextFactory
+    {
+        private readonly SocketTransportOptions _transportOptions;
+        private readonly PipeScheduler _transportScheduler;
+        private readonly ISocketsTrace _trace;
+
+        public SocketConnectionContextFactory(
+            IOptions<SocketTransportOptions> transportOptions,
+            ILoggerFactory loggerFactory)
+        {
+            _transportOptions = transportOptions.Value;
+
+            _transportScheduler = _transportOptions.UnsafePreferInlineScheduling ? PipeScheduler.Inline :
```

Comment:
Since this is important, we should probably just add UnsafePreferInlineScheduling to SocketConnectionOptions. It would be weird to change per socket, but I don't think we should prevent it. Just make sure we still reuse the same IOQueue if it switches back and forth.

Reply:
> Just make sure we still reuse the same IOQueue if it switches back and forth.

I'm not sure I follow what additional guard we need here?  I moved the scheduler to the ConnectionOptions for now, the factory is nice and simple now, it basically just passes everything from the options to SocketConnection

Reply:
The problem with just moving the scheduler is that now people using the SocketConnectionContextFactory can no longer use the IOQueue since it's internal and we very much want them to by default for performance. I guess we could make it public and nullable and use the IOQueue when we normally would if it's null.